libs/providers/flagsmith/src/lib/flagsmith-provider.ts

```diff
@@ -0,0 +1,182 @@
+import {
+  EvaluationContext,
+  FlagValue,
+  JsonValue,
+  Logger,
+  OpenFeatureEventEmitter,
+  Provider,
+  ProviderEvents,
+  ProviderMetadata,
+  ProviderStatus,
+  ResolutionDetails,
+  ResolutionReason,
+} from '@openfeature/web-sdk';
+import { createFlagsmithInstance } from 'flagsmith';
+import { IFlagsmith, IInitConfig, IState } from 'flagsmith/types';
+import { FlagType, typeFactory } from './type-factory';
+
+export class FlagsmithProvider implements Provider {
+  readonly metadata: ProviderMetadata = {
+    name: FlagsmithProvider.name,
+  };
+  //The Flagsmith Client
+  private _client: IFlagsmith;
+  //The Open Feature logger to use
+  private _logger?: Logger;
+  //The status of the provider
+  private _status = ProviderStatus.NOT_READY;
+  //The configuration used for the Flagsmith SDK
+  private _config: IInitConfig;
+  // The Open Feature event emitter
+  events = new OpenFeatureEventEmitter();
+
+  constructor({
+    logger,
+    flagsmithInstance,
+    ...config
+  }: Omit<IInitConfig, 'identity' | 'traits'> & { logger?: Logger; flagsmithInstance?: IFlagsmith }) {
+    this._logger = logger;
+    this._client = flagsmithInstance || createFlagsmithInstance();
+    this._config = config;
+  }
+
+  get status() {
+    return this._status;
+  }
+
+  set status(status: ProviderStatus) {
+    this._status = status;
+  }
+
+  async initialize(context?: EvaluationContext & Partial<IState>) {
+    const identity = context?.targetingKey;
+
+    if (this._client?.initialised) {
+      //Already initialised, set the state based on the new context, allow certain context props to be optional
+      const defaultState = { ...this._client.getState(), identity: undefined, traits: {} };
+      this._client.identity = identity;
+      this._client.setState({
+        ...defaultState,
+        ...(context || {}),
+      });
+      this._status = ProviderStatus.STALE;
+      this.events.emit(ProviderEvents.Stale, { message: 'context has changed' });
+      return this._client.getFlags();
+    }
+
+    const serverState = this._config.state;
+    if (serverState) {
+      this._client.setState(serverState);
+      this.status = ProviderStatus.READY;
+    }
+    return this._client.init({
+      ...this._config,
+      ...context,
+      identity,
+      onChange: (previousFlags, params, loadingState) => {
+        this.status = ProviderStatus.READY;
+        this.events.emit(ProviderEvents.Ready, {
+          message: 'Flags ready',
+        });
+        if (params.flagsChanged) {
+          this.events.emit(ProviderEvents.ConfigurationChanged, {
+            message: 'Flags changed',
+          });
```

Comment:
You can include a list of the flags keys that changed if you have access to that information.

Reply:
Oh great, we do with a slight tweak upstream.